File: Data Structures & Algorithms/products-of-array-discluding-self/submission-4.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        zeros = 0
        product = 1
        for num in nums:
            if num == 0:
                zeros += 1
            else:
                product *= num
        
        if zeros > 1:
            return [0] * n
        
        res = [0] * n
        for i in range(n):
            if zeros > 0:
                if nums[i] == 0:
                    res[i] = product
            else:
                res[i] = product // nums[i]
        
        return res

            
        

File: Data Structures & Algorithms/count-number-of-islands/submission-4.java
class Solution {
    public int numIslands(char[][] grid) {
        int res = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    res += 1;
                    fill(i, j, grid);
                }
            }
        }
        return res;
    }

    public void fill(int r, int c, char[][] grid) {
        if (r < 0 || r >= grid.length || c < 0 || c >= grid[0].length || grid[r][c] == '0') {
            return;
        }

        grid[r][c] = '0';
        fill(r-1, c, grid);
        fill(r+1, c, grid);
        fill(r, c-1, grid);
        fill(r, c+1, grid);
    }
}
